### PR TITLE
Support view change in commitment counter

### PR DESCRIPTION
**Pull request by sergefdrv**
This pull request starts adding view change support. The first component to change is `commitmentCounter`. Please see commit messages for more details.

### CODE REVIEW

File: core/commit.go [hunk 1]
```diff
@@ -142,6 +150,15 @@ func makeCommitmentCounter(f uint32) commitmentCounter {
 		lock.Lock()
 		defer lock.Unlock()
 
+		if prepareView < view {
+			return false, nil
```
Comment (nhoriguchi) [inline core/commit.go:154]: <del>Returning an error string might be useful as other error paths do. </del> Nop, this case might be one that should be silently ignored. I withdraw this comment, sorry.

File: core/commit.go [hunk 2]
```diff
@@ -142,6 +150,15 @@ func makeCommitmentCounter(f uint32) commitmentCounter {
...
+		} else if prepareView > view {
+			view = prepareView
+			firstCV = prepareCV
+			lastDoneCV = 0
```
Comment (nhoriguchi) [inline core/commit.go:158]: I think that `lastDoneCV` should be turned back to the last checkpoint value, but checkpoint is not implemented yet so setting to 0 now. Maybe that's fine.

File: core/commit.go [hunk 3]
```diff
@@ -133,6 +140,7 @@ func makeCommitmentCounter(f uint32) commitmentCounter {
 
 	return func(replicaID uint32, prepare *messages.Prepare) (done bool, err error) {
 		primaryID := prepare.ReplicaID()
+		prepareView := prepare.View()
```
Comment (nhoriguchi) [inline core/commit.go:143]: `primaryID` is always equal to `prepareView % n`, but if that's not the case, the message should be suspected seriously. So is it worth asserting that?

File: core/commit_test.go [hunk 4]
```diff
@@ -237,6 +238,40 @@ func TestMakeCommitmentCounter(t *testing.T) {
 			replicaID: 2,
 			ok:        false,
 			done:      false,
+		}, {
+			desc:      "Fist commitment in a new view",
```
Comment (nhoriguchi) [inline core/commit_test.go]: s/Fist/First/ ?
Ditto line 269.

File: core/commit.go [hunk 5]
```diff
@@ -125,6 +125,13 @@ func makeCommitmentCounter(f uint32) commitmentCounter {
 	var (
 		lock sync.Mutex
 
+		// Current view number
+		view uint64
```
Comment (ynamiki) [inline core/commit.go:129]: At line 153, this variable is used without explicit initialization. Do we add ` = 0` here to clarify we expect zero for this?

File: core/commit.go (hunk 2 above)
Comment (sergefdrv) [reply core/commit.go:158]: The only purpose of `lastDoneCV` is to limit `prepareStates` in size. In fact, `lastDoneCV` is the highest counter value that is known to have reached the quorum. Zero is not a valid counter value, so it indicates that no quorum has been reached in the view.

File: core/commit.go (hunk 3 above)
Comment (sergefdrv) [reply core/commit.go:143]: As documented in definition of `commitmentCounter` type, the passed Prepare message is supposed to be validated for internal consistency. This means `prepare.ReplicaID()` must denote the primary replica of view `prepare.View()` in the current configuration.

**Comment by sergefdrv** [comment]
Summary of [changes](https://github.com/hyperledger-labs/minbft/compare/2ee32e02548becd6590b258c713c3f19fba8891b..0a72bb25260e6e8ace4b5ec902b4a93b032ccdeb):

-  Fixed typos

File: core/commit.go (hunk 5 above)
Comment (sergefdrv) [reply core/commit.go:129]: It is not really necessary. Unlike C/C++, Go variables are always initialized with [a default value](https://golang.org/ref/spec#The_zero_value), which is `0` for numeric types.

File: core/commit.go (hunk 2 above)
Comment (nhoriguchi) [reply core/commit.go:158]: Okay, thanks for clarification.

**Review by nhoriguchi — APPROVED** [review]
The proposed change looks fine to me. Thanks.

File: core/commit.go (hunk 5 above)
Comment (ynamiki) [reply core/commit.go:129]: Yes, however I tend to explicitly initialize a variable if it is used without assignment. It's my coding preference so no problem.

**Review by ynamiki — APPROVED** [review]
Looks good to me, thanks.

**Pull request merged**